### PR TITLE
[python] More gracefully handle invalid operators in query-condition parser

### PR DESCRIPTION
**Pull request by johnkerl**
**Issue and/or context:** [[sc-51255]](https://app.shortcut.com/tiledb-inc/story/51255) 

**Changes:** Provide user feedback on operators which are (a) valid in Python, but we don't accept; (b) invalid in Python.

**Notes for Reviewer:**

Examples:

```
#!/usr/bin/env python
import tiledbsoma
import sys
expr = 'foo is True'
if len(sys.argv) == 2:
    expr = sys.argv[1]
with tiledbsoma.Experiment.open('/var/p') as exp: # /var/p is a copy of pbmc3k; any Experiment will do
    exp.obs.read(value_filter=expr).concat()
```
```
johnkerl@exade[prod][main][python]$ python q1.py 'foo &&& True'
Traceback (most recent call last):
  File "/Users/johnkerl/git/single-cell-data/TileDB-SOMA/apis/python/src/tiledbsoma/_query_condition.py", line 115, in __attrs_post_init__
    self.tree = ast.parse(self.expression, mode="eval")
  File "/usr/local/Cellar/python@3.9/3.9.16/Frameworks/Python.framework/Versions/3.9/lib/python3.9/ast.py", line 50, in parse
    return compile(source, filename, mode, flags,
  File "<unknown>", line 1
    foo &&& True
         ^
SyntaxError: invalid syntax

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/johnkerl/git/single-cell-data/TileDB-SOMA/apis/python/q1.py", line 12, in <module>
    exp.obs.read(value_filter=expr).concat()
  File "/Users/johnkerl/git/single-cell-data/TileDB-SOMA/apis/python/src/tiledbsoma/_dataframe.py", line 411, in read
    sr.set_condition(QueryCondition(value_filter), handle.schema)
  File "<attrs generated init tiledbsoma._query_condition.QueryCondition>", line 3, in __init__
    self.__attrs_post_init__()
  File "/Users/johnkerl/git/single-cell-data/TileDB-SOMA/apis/python/src/tiledbsoma/_query_condition.py", line 117, in __attrs_post_init__
    raise SOMAError(
tiledbsoma._exception.SOMAError: Could not parse the given QueryCondition statement: foo &&& True
```

```
johnkerl@exade[prod][kerl/qc-parser-bad-operator][python]$ python q1.py 'foo is True'
Traceback (most recent call last):
  File "/Users/johnkerl/git/single-cell-data/TileDB-SOMA/apis/python/q1.py", line 12, in <module>
    exp.obs.read(value_filter=expr).concat()
  File "/Users/johnkerl/git/single-cell-data/TileDB-SOMA/apis/python/src/tiledbsoma/_dataframe.py", line 411, in read
    sr.set_condition(QueryCondition(value_filter), handle.schema)
tiledbsoma._exception.SOMAError: SOMAError: the `is` operator is not supported

At:
  /Users/johnkerl/git/single-cell-data/TileDB-SOMA/apis/python/src/tiledbsoma/_query_condition.py(137): init_query_condition
  /Users/johnkerl/git/single-cell-data/TileDB-SOMA/apis/python/src/tiledbsoma/_dataframe.py(411): read
  /Users/johnkerl/git/single-cell-data/TileDB-SOMA/apis/python/q1.py(12): <module>
```

```
johnkerl@exade[prod][kerl/qc-parser-bad-operator][python]$ python q1.py 'foo is not True'
Traceback (most recent call last):
  File "/Users/johnkerl/git/single-cell-data/TileDB-SOMA/apis/python/q1.py", line 12, in <module>
    exp.obs.read(value_filter=expr).concat()
  File "/Users/johnkerl/git/single-cell-data/TileDB-SOMA/apis/python/src/tiledbsoma/_dataframe.py", line 411, in read
    sr.set_condition(QueryCondition(value_filter), handle.schema)
tiledbsoma._exception.SOMAError: SOMAError: the `is not` operator is not supported

At:
  /Users/johnkerl/git/single-cell-data/TileDB-SOMA/apis/python/src/tiledbsoma/_query_condition.py(137): init_query_condition
  /Users/johnkerl/git/single-cell-data/TileDB-SOMA/apis/python/src/tiledbsoma/_dataframe.py(411): read
  /Users/johnkerl/git/single-cell-data/TileDB-SOMA/apis/python/q1.py(12): <module>
```

### CODE REVIEW

**Review by ryan-williams — APPROVED** [review]
lgtm, I was trying to figure out whether there's a way to exercise the new `else:` branch on [L257](https://github.com/single-cell-data/TileDB-SOMA/pull/2830/files#diff-5b57a212c31c7b2a1e2fd4ca307632427776256adf1f6eca2a9eeeb048cfecd4R257), but I couldn't find one.

I also would like to verify the error messages in some of those `pytest.raises(SOMAError)` contexts, here's an example https://github.com/single-cell-data/TileDB-SOMA/commit/b28a270c75355de800f514bc486f1e59d7279118, feel free to take or leave.

There are simple ops that still raise the non-descriptive `'Name' object has no attribute 'combine'` SOMAError (here and in the TileDB-Py code this was copied from.

**Comment by johnkerl** [comment]
> There are simple ops that still raise the non-descriptive `'Name' object has no attribute 'combine'` SOMAError

@ryan-williams thanks -- please do share your examples!

**Comment by johnkerl** [comment]
@ryan-williams when you do share your examples I'll be able to check to see if they're handled by also-open #2831

**Comment by johnkerl** [comment]
Rebased in #2831 which has since been merged

File: apis/python/tests/test_query_condition.py [hunk 1]
```diff
@@ -226,18 +226,36 @@ def test_eval_error_conditions(malformed_condition):
...
-    "text_and_exception",
+    "expression_and_message",
     [
-        ["louvain == leukocyte", SOMAError],
-        ["louvain == leuko-cyte", SOMAError],  # minus sign looks like an operator
+        # ["foo is True", SOMAError],
+        # ["foo is not True", SOMAError],
+        # ["foo &&& bar", SOMAError],
+        # ["louvain == leukocyte", SOMAError],
+        # ["louvain == leuko-cyte", SOMAError],  # minus sign looks like an operator
```
Comment (ryan-williams) [inline apis/python/tests/test_query_condition.py:235]: remove?

Comment (johnkerl) [reply apis/python/tests/test_query_condition.py]: derpthx

**Comment by johnkerl** [comment]
I'll take follow-ups (@ryan-williams mentions `a | b`) on a separate PR

**Pull request merged**